README.md
# Tock.js

A nice and accurate Javascript count-up/countdown timer.

Based on [Tock](https://github.com/mrchimp/Tock) by Mr Chimp and [an idea by James Edwards](http://www.sitepoint.com/creating-accurate-timers-in-javascript/).

License: [MIT](https://github.com/Falc/Tock.js/blob/master/LICENSE)

## Features

* Pure Javascript, no dependencies.
* Self-correcting time based on the system clock.
* Count-up/countdown.
* Custom intervals and starting times.
* Start, stop, resume and reset.
* Callbacks: onTick(), onComplete(), onStart(), onStop() and onReset().
* Flexible time formatting.

## The problem with setInterval/setTimeout

Javascript's iteration timers using `setInterval()` or `setTimeout()` are not realiable enough. They create a difference between the iterated time and the real time.
That difference can be compensated by using the actual time to adjust the timer.

These examples made by James Edwards illustrates the differences: http://www.sitepoint.com/examples/timeradjust/examples.html

## Usage

**Tock.js** is not a full widget for a website, it is a timer library. You can use it either to display the result on screen or to make things work in the backend.

Let's see an example:

### 1) Some HTML
```html
<span id="time">0</span>
<button id="start">Start</button>
<button id="stop">Stop</button>
<button id="reset">Reset</button>
```

### 2) Tock instance
```js
var timer = new Tock(options);
```

The `options` parameter allows to customize the timer easily. All of them are optional, of course.
```js
var options = {
  countdown: true,
  startTime: 10000
  interval: 15,
  onTick: function() {/* Some code... */},
  onComplete: function() {/* Some code... */},
  onStart: function() {/* Some code... */},
  onStop: function() {/* Some code... */},
  onReset: function() {/* Some code... */}
}
```

#### Options

* **countdown**: *boolean*
  * If true, the timer will count down. Otherwise, it will count up.
  * Default: false.
* **startTime**: *integer*
  * The starting time in milliseconds. This should be set to a number greater than 0 when `countdown` is true.
  * Default: 0.
* **interval**: *integer*
  * The interval in milliseconds. This defines how often the timer will tick.
  * Default: 0.
* **onTick**: *function*
  * A function that will be called on every tick.
  * Default: A mockup function that will suggest to set a real one.
* **onComplete**: *function*
  * A function that will be called when a countdown reaches zero.
  * Default: A mockup function that will suggest to set a real one.
* **onStart**: *function*
  * A function that will be called when a timer is started.
  * Default: Dummy function.
* **onStop**: *function*
  * A function that will be called when a timer is stopped.
  * Default: Dummy function.
* **onReset**: *function*
  * A function that will be called when a timer is reset.
  * Default: Dummy function.

#### Methods

* **start()**
  * Starts the timer.
* **stop()**
  * Stops the timer. It is possible to resume the timer calling `start()` again.
* **reset()**
  * Resets the timer to `startTime`.
* **lap(format)**
  * Gets the current time in milliseconds.
  * The `format` parameter is optional, it is passed to the `format()` method.
* **format(time, format)**
  * Returns a `time` (milliseconds) in the specified `format`. (See the [Time Format](#time-format) section)

#### Callbacks

On every tick (every `interval` milliseconds), `onTick()` will be called.
```js
var options = {
  // some options...
  onTick: function() {
    // Pure javascript
    document.querySelector('#time').innerHTML = timer.lap();

    // Using jQuery/Zepto
    $('#time').text(timer.lap());
  }
}
```

When a countdown timer reaches zero, `onComplete()` will be called.
```js
var options = {
  // some options...
  onComplete: function() {
    alert("Countdown complete!");
  }
}
```

When a timer starts, `onStart()` will be called.
```js
var options = {
  // some options...
  onStart: function() {
    console.log("Start!");
  }
}
```
When a timer stop, `onStop()` will be called.
```js
var options = {
  // some options...
  onStop: function() {
    console.log("Stop!");
  }
}
```

When a timer is reset, `onReset()` will be called.
```js
var options = {
  // some options...
  onReset: function() {
    console.log("Reset!");
  }
}
```

### 3) Add some controls

The timer can be controlled easily with some buttons and events.

Start button:
```js
// Pure Javascript
document.querySelector('#start').onclick = function() {
  timer.start();
});

// Using jQuery/Zepto
$('#start').on('click', function() {
  timer.start();
});
```

Stop button:
```js
// Pure Javascript
document.querySelector('#stop').onclick = function() {
  timer.stop();
});

// Using jQuery/Zepto
$('#stop').on('click', function() {
  timer.stop();
});
```

Reset button:
```js
// Pure Javascript
document.querySelector('#reset').onclick = function() {
  timer.reset();
});

// Using jQuery/Zepto
$('#reset').on('click', function() {
  timer.reset();
});
```

It is possible to set a "lap" button too:
```js
// Pure Javascript
document.querySelector('#lap').onclick = function() {
  document.querySelector('#time').innerHTML += '<br />' + timer.lap();
});

// Using jQuery/Zepto
$('#lap').on('click', function() {
  $('#time').append('<br />' + timer.lap());
});
```

## Time Format

The `format(time, format)` method allows to customize the `format` of the given `time` (passed as milliseconds). It will replace the placeholders found in the `format` parameter, with the corresponding time information:

* **{H}**: Hours
* **{h}**: Hours (same as `{H}`)
* **{M}**: Minutes
* **{m}**: Minutes (0 - 59)
* **{S}**: Seconds
* **{s}**: Seconds (0 - 59)
* **{L}**: Milliseconds (Default)
* **{l}**: Milliseconds (0 - 999)

Capital letters are raw values. This means that `{M}` displays the full number of minutes (i.e. 127 minutes), but `{m}` displays only the number of minutes elapsed since the last hour (7 in 127).

Double letters are used to retrieve the time with leading zeros. If `{m}` returns **7**, `{mm}` will return **07**.

Examples:
```php
    Time: 3877012 (1 hour, 4 minutes, 37 seconds and 12 milliseconds)
    '{hh}:{mm}:{ss}.{ll}'        // 01:04:37.012
    '{MM}'                       // 64
    '{SS}.{ll}'                  // 3877.012
    '{MM} min and {ss} s'        // 64 min and 37 s
    '{h}:{m} vs {hh}:{mm}'       // 1:4 vs 01:04
```


CHANGELOG.md
# Changelog

## 2.1.0 (2019-12-17)

* Adapt to UMD.
* Organize files.
* Replace bower.json with package.json.

## 2.0.0 (2014-11-22)

 * Add new callbacks: onStart(), onStop() and onReset().
 * Add Date.now polyfill for IE8 support.
 * Rename options (breaks backward compatibility).
 * Add gruntfile.
 * Refactoring.

## 1.2.1 (2014-11-09)

 * Print the most updated time when calling stop().
 * Minor fixes related to callbacks.
 * Refactoring.

## 1.2.0 (2014-11-07)

 * Ensure reset() works as expected.
 * Add bower.json.

## 1.1.0 (2014-03-11)

 * Fix a critical bug in countdowns.
 * Better examples.

## 1.0.0 (2013-09-17)

 * Initial release.


Gruntfile.js
module.exports = function(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    app: {
      name:     'Tock.js',
      version:  '<%= pkg.version %>',
      license:  '<%= pkg.license %>',
      homepage: '<%= pkg.homepage %>',
      paths: {
        src:  'src',
        dist: 'dist'
      },
      files: {
        gruntfile: ['Gruntfile.js'],
        js:        ['tock.js']
      }
    },
    jshint: {
      options: {
        camelcase:  true,
        curly:      true,
        eqeqeq:     true,
        jquery:     true,
        newcap:     true,
        undef:      true,
        unused:     true,
        validthis:  true
      },
      gruntfile: {
        options: {
          predef: [
            'module'
          ]
        },
        src: '<%= app.files.gruntfile %>'
      },
      js: {
        options: {
          browser: true,
          devel:   true,
          predef:  [
            'define',
            'module',
            'require'
          ]
        },
        src: '<%= app.paths.src %>/<%= app.files.js %>'
      }
    },
    jscs: {
      options: {
        'disallowSpacesInNamedFunctionExpression': {
          'beforeOpeningRoundBrace': true
        },
        'disallowSpacesInFunctionExpression': {
          'beforeOpeningRoundBrace': true
        },
        'disallowSpacesInAnonymousFunctionExpression': {
          'beforeOpeningRoundBrace': true
        },
        'disallowSpacesInFunctionDeclaration': {
          'beforeOpeningRoundBrace': true
        },
        'disallowSpaceBeforeBinaryOperators': [
          ','
        ],
        'requireSpaceBeforeBinaryOperators': true,
        'requireSpaceAfterBinaryOperators': true,
        'requireCamelCaseOrUpperCaseIdentifiers': true,
        'requireCapitalizedConstructors': true,
        'disallowMixedSpacesAndTabs': true,
        'disallowTrailingWhitespace': true,
        'disallowTrailingComma': true,
        'disallowSpaceAfterPrefixUnaryOperators': true,
        'disallowSpaceBeforePostfixUnaryOperators': true,
        'disallowSpacesInsideArrayBrackets': true,
        'disallowSpacesInsideParentheses': true,
        'requireSpaceAfterKeywords': [
          'if',
          'else',
          'for',
          'while',
          'do',
          'switch',
          'case',
          'return',
          'try',
          'catch',
          'typeof'
        ],
        'validateIndentation': 2,
        'validateLineBreaks': 'LF',
        'validateQuoteMarks': true
      },
      gruntfile: {
        options: {
          'requireCamelCaseOrUpperCaseIdentifiers': 'ignoreProperties'
        },
        src: '<%= app.files.gruntfile %>'
      },
      js: {
        src: '<%= app.paths.src %>/<%= app.files.js %>'
      }
    },
    umd: {
      js: {
        src:            '<%= app.paths.src %>/<%= app.files.js %>',
        dest:           '<%= app.paths.dist %>/<%= app.files.js %>',
        amdModuleId:    '<%= pkg.name %>',  // AMD
        objectToExport: '<%= app.name %>',  // CommonJS
        globalAlias:    '<%= app.name %>',  // Global
        deps:           {}
      }
    },
    uglify: {
      options: {
        banner: '// <%= app.name %> (version <%= app.version %>) <%= app.homepage %>\n// License: <%= app.license %>\n',
        compress: {
          comparisons: false
        }
      },
      js: {
        files: [{
          src:    '<%= umd.js.dest %>',
          ext:    '.min.js',
          expand: true
        }]
      }
    },
    watch: {
      gruntfile: {
        files: '<%= app.files.gruntfile %>',
        tasks: ['check:gruntfile', 'build']
      },
      js: {
        options: {
          cwd: '<%= app.paths.src %>'
        },
        files: '<%= app.files.js %>',
        tasks: ['build:js']
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-jscs');
  grunt.loadNpmTasks('grunt-notify');
  grunt.loadNpmTasks('grunt-umd');

  grunt.registerTask('check:gruntfile', ['jshint:gruntfile', 'jscs:gruntfile']);
  grunt.registerTask('check:js', ['jshint:js', 'jscs:js']);
  grunt.registerTask('check', ['check:gruntfile', 'check:js']);

  grunt.registerTask('build:js', ['check:js', 'umd:js', 'uglify:js']);
  grunt.registerTask('build', ['build:js']);

  grunt.registerTask('default', ['check:gruntfile', 'build']);
};


src/tock.js
/**
 * Tock.js
 *
 * A nice and accurate Javascript count-up/countdown timer.
 *
 * Based on:
 * - Tock by Mr Chimp: https://github.com/mrchimp/Tock
 * - An idea by James Edwards: http://www.sitepoint.com/creating-accurate-timers-in-javascript/
 *
 * Authors:
 * - Aitor García (aitor.falc@gmail.com)
 * - Roberto Salicio
 *
 * Version: 2.1.0
 * License: MIT (https://github.com/Falc/Tock.js/blob/master/LICENSE)
 */

// Date.now polyfill for IE<9
Date.now = Date.now || function() {return +new Date();};

var Tock = (function() { // jshint ignore:line
  /**
   * Constructor.
   */
  function Tock(options) {
    options = options || {};

    // Default params
    this.isCountdown = options.countdown || false;
    this.startTime = options.startTime || 0;
    this.interval = options.interval || 10;
    this.onTick = options.onTick || function() {console.warn('Callback function for "onTick" is not defined.');};
    this.onComplete = options.onComplete || function() {console.warn('Callback function for "onComplete" is not defined.');};
    this.onStart = options.onStart || function() {};
    this.onStop = options.onStop || function() {};
    this.onReset = options.onReset || function() {};

    this.isRunning = false;
    this.timeout = null;

    this.stoppedTime = this.startTime;
    this.startedAt = 0;
    this.time = 0;
    this.elapsed = '0.0';
  }

  /**
  * Starts the timer.
  */
  Tock.prototype.start = function() {
    // Don't try to start if it is already running
    if (this.isRunning) {
      return;
    }

    // A countdown that has finished should be reseted before starting again
    if (this.isCountdown && this.stoppedTime <= 0) {
      return;
    }

    var me = this;

    this.startedAt = Date.now();
    this.time = 0;
    this.elapsed = '0.0';
    this.isRunning = true;

    this.timeout = window.setTimeout(
      function() {me.tick();},
      this.interval
    );

    if (typeof this.onStart === 'function') {
      this.onStart();
    }
  };

  /**
   * Stops the timer.
   */
  Tock.prototype.stop = function() {
    // Don't try to stop if it is not running
    if (!this.isRunning) {
      return;
    }

    // Keep the lap time before stopping
    this.stoppedTime = this.lap();
    this.isRunning = false;

    window.clearTimeout(this.timeout);

    if (typeof this.onTick === 'function') {
      this.onTick();
    }

    if (typeof this.onStop === 'function') {
      this.onStop();
    }
  };

  /**
   * Resets the timer.
   */
  Tock.prototype.reset = function() {
    this.isRunning = false;
    window.clearTimeout(this.timeout);
    this.timeout = null;

    this.stoppedTime = this.startTime;
    this.startedAt = 0;
    this.time = 0;
    this.elapsed = '0.0';

    if (typeof this.onTick === 'function') {
      this.onTick();
    }

    if (typeof this.onReset === 'function') {
      this.onReset();
    }
  };

  /**
   * Tick.
   *
   * Called every "this.interval" seconds.
   */
  Tock.prototype.tick = function() {
    var me = this;

    this.time += this.interval;

    this.elapsed = Math.floor(this.time / this.interval) / 10;

    if (Math.round(this.elapsed) === this.elapsed) {
      this.elapsed += '.0';
    }

    var diff = (Date.now() - this.startedAt) - this.time;

    var nextIntervalIn = this.interval - diff;

    if (typeof this.onTick === 'function') {
      this.onTick();
    }

    if (this.isCountdown && (this.stoppedTime - this.time < 0)) {
      this.stoppedTime = 0;
      this.isRunning = false;

      if (typeof this.onComplete === 'function') {
        this.onComplete();
      }
    }

    if (nextIntervalIn <= 0) {
      var missedTicks = Math.floor(Math.abs(nextIntervalIn) / this.interval);
      this.time += missedTicks * this.interval;

      if (this.isRunning) {
        this.tick();
      }

      return;
    }

    if (this.isRunning) {
      this.timeout = window.setTimeout(
        function() {me.tick();},
        nextIntervalIn
      );
    }
  };

  /**
   * Gets the current time in the specified format (default is milliseconds).
   */
  Tock.prototype.lap = function(format) {
    if (!this.isRunning) {
      // Initial status
      if (this.startedAt === 0) {
        return this.format(this.startTime, format);
      }

      return this.format(this.stoppedTime, format);
    }

    if (this.isCountdown) {
      var lapTime = this.stoppedTime - (Date.now() - this.startedAt);
      if (lapTime < 0) {
        lapTime = 0;
      }

      return this.format(lapTime, format);
    }

    return this.format(this.stoppedTime + (Date.now() - this.startedAt), format);
  };

  /**
   * Returns a time (in milliseconds) formatted as specified.
   *
   * The format is passed as a string and can contain placeholders that will be replaced.
   * Double letters are used for leading zeros.
   * Capital letters are raw values.
   *
   * {H}: Hours
   * {h}: Hours (same as H)
   * {M}: Minutes
   * {m}: Minutes (0 - 59)
   * {S}: Seconds
   * {s}: Seconds (0 - 59)
   * {L}: Milliseconds (Default)
   * {l}: Milliseconds (0 - 999)
   *
   * Example: 3877012 (1 hour, 4 minutes, 37 seconds and 12 milliseconds)
   * '{hh}:{mm}:{ss}.{ll}'        => 01:04:37.012
   * '{MM}'                       => 64
   * '{SS}.{ll}'                  => 3877.012
   * '{MM} min and {ss} s'        => 64 min and 37 s
   * '{h}:{m} vs {hh}:{mm}'       => 1:4 vs 01:04
   */
  Tock.prototype.format = function(time, format) {
    if (format === undefined || format === '{L}') {
      return time;
    }

    var formatted = format;

    var info = {};
    var x = time;

    var milliseconds = x % 1000;
    info.L = x.toString();
    info.LL = '000'.substring(0, 3 - x.toString().length) + x.toString();
    info.l = milliseconds.toString();
    info.ll = '000'.substring(0, 3 - milliseconds.toString().length) + milliseconds.toString();

    x = (x - milliseconds) / 1000;
    var seconds = x % 60;
    info.S = x.toString();
    info.SS = '00'.substring(0, 2 - x.toString().length) + x.toString();
    info.s = seconds.toString();
    info.ss = '00'.substring(0, 2 - seconds.toString().length) + seconds.toString();

    x = (x - seconds) / 60;
    var minutes = x % 60;
    info.M = x.toString();
    info.MM = '00'.substring(0, 2 - x.toString().length) + x.toString();
    info.m = minutes.toString();
    info.mm = '00'.substring(0, 2 - minutes.toString().length) + minutes.toString();

    var hours = (x - minutes) / 60;
    info.H = hours.toString();
    info.HH = '00'.substring(0, 2 - hours.toString().length) + hours.toString();
    info.h = info.H;
    info.hh = info.h;

    var letters = ['HH', 'H', 'hh', 'h', 'MM', 'M', 'mm', 'm', 'SS', 'S', 'ss', 's', 'LL', 'L', 'll', 'l'];

    for (var i = letters.length - 1; i >= 0; i--) {
      formatted = formatted.replace(new RegExp('{' + letters[i] + '}', 'g'), info[letters[i]]);
    }

    return formatted;
  };

  return Tock;
}());
